sound/soc/sof/intel/hda-loader.c

```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+static int hda_dsp_boot_imr(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	const struct sof_intel_dsp_desc *chip = hda->desc;
+	int ret, i;
+
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
+
+	/* power up & unstall/run core 0 to run the firmware */
+	ret = hda_dsp_enable_core(sdev, BIT(0));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: dsp core start failed %d\n", ret);
+		return -EIO;
+	}
+
+	/* DSP is powered up, set all SSPs to slave mode */
+	for (i = 0; i < chip->ssp_count; i++) {
+		snd_sof_dsp_update_bits_unlocked(sdev, HDA_DSP_BAR,
+						 chip->ssp_base_offset
+						 + i * SSP_DEV_MEM_SIZE
+						 + SSP_SSC1_OFFSET,
+						 SSP_SET_SLAVE,
+						 SSP_SET_SLAVE);
+	}
+
+	/* enable IPC interrupts */
+	hda_dsp_ipc_int_enable(sdev);
+
+	/*
+	 * When a SoundWire link is in clock stop state, a Slave
+	 * device may trigger in-band wakes for events such as jack
+	 * insertion or acoustic event detection. This event will lead
+	 * to a WAKEEN interrupt, handled by the PCI device and routed
+	 * to PME if the PCI device is in D3. The resume function in
+	 * audio PCI driver will be invoked by ACPI for PME event and
+	 * initialize the device and process WAKEEN interrupt.
+	 *
+	 * The WAKEEN interrupt should be processed ASAP to prevent an
+	 * interrupt flood, otherwise other interrupts, such IPC,
+	 * cannot work normally.  The WAKEEN is handled after the ROM
+	 * is initialized successfully, which ensures power rails are
+	 * enabled before accessing the SoundWire SHIM registers
+	 */
+	if (!sdev->first_boot)
+		hda_sdw_process_wakeen(sdev);
```

Comment:
lines 354..383 look like copy/paste. can we avoid this please? At the very least introduce helpers for SSP and SoundWire.

Reply:
> lines 354..383 look like copy/paste. can we avoid this please? At the very least introduce helpers for SSP and SoundWire.

Um, I thought about this, then I didn't want to change the sequence (e.g. to move SSP/SDW to the last) to introduce unexpected risk and that's why you see this...

Since you are confirming the SSP/SDW can be moved to the very last, let me share that part to avoid copy/paste here.

Reply:
I am not confirming anything wrt SSP/SDW, you need to keep the same order as before, but call a helper. Having duplicated code and comments isn't good.

Reply:
> I am not confirming anything wrt SSP/SDW, you need to keep the same order as before, but call a helper. Having duplicated code and comments isn't good.

got you.